cli/state.rs

```diff
@@ -357,22 +357,43 @@ impl ModuleLoader for State {
   fn prepare_load(
     &self,
     _load_id: ModuleLoadId,
-    _module_specifier: &ModuleSpecifier,
-    _maybe_referrer: Option<String>,
-    _is_dyn_import: bool,
+    module_specifier: &ModuleSpecifier,
+    maybe_referrer: Option<String>,
+    is_dyn_import: bool,
   ) -> Pin<Box<dyn Future<Output = Result<(), ErrBox>>>> {
-    // TODO(bartlomieju):
-    // 1. recursively:
-    //    a) resolve specifier
-    //    b) check permission if dynamic import
-    //    c) fetch/download source code
-    //    d) parse the source code and extract all import/exports (dependencies)
-    //    e) add discovered deps and loop algorithm until no new dependencies
-    //        are discovered
-    // 2. run through appropriate compiler giving it access only to
-    //     discovered files
-
-    async { Ok(()) }.boxed_local()
+    let module_specifier = module_specifier.clone();
+    let state = self.borrow();
+    let target_lib = state.target_lib.clone();
+    let maybe_import_map = state.import_map.clone();
+    let permissions = if state.is_main && !is_dyn_import {
+      Permissions::allow_all()
+    } else {
+      state.permissions.clone()
+    };
+    let global_state = state.global_state.clone();
```

Comment:
Implementation of this method requires to solve issue described in https://github.com/denoland/deno/issues/1692 as well. Currently TS compiler is spun up during `fetch_compiled_module()` call - if root module was already compiled and hasn't changed then cached code will be used. If one of root's dependencies has changed then TS compiler will be spun up. Implementing `prepare_load()` means that we need to manually handle outdated dependencies - because this function will be the only place that can spin up TS compiler; `load()` will only access already compiled source code in `$DENO_DIR`.

CC @kitsonk @ry

Reply:
Added basic handling of cache invalidation by saving dependency list in "module graph" meta file.